### PR TITLE
Fix crash on hipblaslt algo selection. Add test precision workaround …

### PR DESCRIPTION
**Pull request by ipanfilo**
…on MI308. Add pytorch 2.5 support

# Description

-Fix crash in corner cases with hipBlasLt algo selection
-Guard selected hipblaslt algo logging with NVTE_LOG_GEMM_CONFIG
-Add workaround for numerical test error on MI308 with ROCm6.3
-Add torch.compiler reset to avoid test failure on pytorch 2.5 due to cache limit reaching

Fixes # (issue)

https://github.com/ROCm/frameworks-internal/issues/10435
https://github.com/ROCm/frameworks-internal/issues/10561
Crash on debugging https://github.com/ROCm/frameworks-internal/issues/9693

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

### CODE REVIEW

File: transformer_engine/common/gemm/rocm_gemm.cu [hunk 1]
```diff
@@ -1307,6 +1304,11 @@ void hipblaslt_gemm(const Tensor *inputA,
       cached_algo.ws_size_min = algoArr[bestAlgo].workspaceSize;
       cached_algo.ws_size_max = workspaceSize;
 
+      if (const char* env_p = std::getenv("NVTE_LOG_GEMM_CONFIG") ) {
```
Comment (wenchenvincent) [inline transformer_engine/common/gemm/rocm_gemm.cu]: `NVTE_LOG_GEMM_CONFIG` is used to log the gemm configs. And this is for log the solution information in autotuning. Let's use a different env var like `NVTE_LOG_HIPBLASLT_TUNING`

File: transformer_engine/pytorch/distributed.py [hunk 2]
```diff
@@ -242,14 +242,14 @@ def _get_active_autocast_contexts():
 
     gpu_autocast_enabled = torch.is_autocast_enabled()
     gpu_autocast_dtype = torch.get_autocast_gpu_dtype()
-    gpu_autocast_ctx = torch.cuda.amp.autocast(
-        gpu_autocast_enabled, gpu_autocast_dtype, autocast_cached
+    gpu_autocast_ctx = torch.amp.autocast('cuda',
```
Comment (wenchenvincent) [inline transformer_engine/pytorch/distributed.py:245]: Is this change for pytorch 2.5?

File: tests/pytorch/test_numerics.py [hunk 3]
```diff
@@ -670,6 +670,7 @@ def test_gpt_full_activation_recompute(dtype, bs, model, fp8, fp8_model_params,
         pytest.skip(reason_for_no_fp8)
 
     config = model_configs[model]
+    torch.compiler.reset() # avoid cache size limit overflow
```
Comment (wenchenvincent) [inline tests/pytorch/test_numerics.py:673]: Is this the workaround for the inductor issue? If so, might need to put a TODO here to remove it later when the issue is fixed in pytorch.

File: transformer_engine/pytorch/distributed.py (hunk 2 above)
Comment (ipanfilo) [reply transformer_engine/pytorch/distributed.py:245]: Yes, old API is deprecated. It is not removed yet, just throws warnings

File: tests/pytorch/test_numerics.py (hunk 3 above)
Comment (ipanfilo) [reply tests/pytorch/test_numerics.py:673]: It is workaround for new JIT cache overflow handling like described in https://github.com/ROCm/frameworks-internal/issues/10561

**Review by wangye805 — APPROVED** [review]
LGTM
